src/java/main/org/apache/zookeeper/server/NIOServerCnxnFactory.java

```diff
@@ -842,7 +868,26 @@ private void addCnxn(NIOServerCnxn cnxn) {
 
     protected NIOServerCnxn createConnection(SocketChannel sock,
             SelectionKey sk, SelectorThread selectorThread) throws IOException {
-        return new NIOServerCnxn(zkServer, sock, sk, this, selectorThread);
+
+        NIOServerCnxn cnxn = null;
+
+        if (serverCnxnClassCtr != null) {
```

Comment:
why not set serverCnxnClassCtr to NIOServerCnxn by default to avoid looking up the right constructor every time we want to create a new connection (a bit of a hot path)?